Add and use MSBuildTarget as opposed to PlatformTarget

marcosnz opened the pull request:
Issue #522

patriksvensson commented:
@marcosnz 

We actually want to take `PlatformTarget` into consideration as well.

I was thinking that we introduce a `MSBuildPlatform` that we check.
This so we can infer what architecture of MSBuild to use simply from the 
platform that should be built. If there's a special case (such as the Silverlight one), 
the user can override by specifying a specific `MSBuildPlatform`.

```csharp
public enum MSBuildPlatform
{
  // Autodetect platform from platfgorm target and environment
  Auto = 0,
  // Explicitly use x86
  x86 = 1,
  // Explicitly use x64
  x64 = 2
}
```

```csharp
private static DirectoryPath GetVisualStudioPath(ICakeEnvironment environment, PlatformTarget target, MSBuildPlatform platform, string version)
{
    // Get the bin path (default to x86).
    var programFilesPath = environment.GetSpecialPath(SpecialPath.ProgramFilesX86);
    var binPath = programFilesPath.Combine(string.Concat("MSBuild/", version, "/Bin"));

    if(platform == MSBuildPlatform.Auto)
    {
      // We want the platform target to decide
      // what MSBuild architecture to use.
      
      if (target == PlatformTarget.MSIL)
      {
          if (environment.Is64BitOperativeSystem())
          {
              binPath = binPath.Combine("amd64");
          }
      }
      if (target == PlatformTarget.x64)
      {
          binPath = binPath.Combine("amd64");
      }
    }
    else
    {
      // We want to use a specific MSBuild architecture.
      
      if (platform == MSBuildPlatform.x64)
      {
          binPath = binPath.Combine("amd64");
      }
    }
    
    return binPath;
}
```

@gep13 @devlead Any more thoughts about this?

marcosnz commented:
What about we set the MSBuildTarget based on PlatformTarget before passing to MSBuildReolver?
{code}
            MSBuildTarget target = settings.MSBuildTarget ?? MSBuildTarget.Default;

            if (target == MSBuildTarget.Default)
            {
                switch (settings.PlatformTarget)
                {
                    case PlatformTarget.x86:
                        target = MSBuildTarget.x86;
                        break;
                    case PlatformTarget.x64:
                        target = MSBuildTarget.x64;
                        break;
                }
            }

         var path = MSBuildResolver.GetMSBuildPath(_fileSystem, _environment, settings.ToolVersion, target);
  {code}

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildTarget.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// Represents an MSBuild exe target.
+    /// </summary>
+    public enum MSBuildTarget
```
Should be renamed to `MSBuildPlatform` to avoid confusion with [MSBuild targets](https://msdn.microsoft.com/en-us/library/ms171462.aspx). Will also follow Team Builds approach to this.

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildTarget.cs:
```diff
@@ -0,0 +1,25 @@
+        /// <summary>
+        /// Will build using MSBuild version based on PlatformTarget/Host OS.
+        /// </summary>
+        Default = 0,
```
Would prefer this to be called `Automatic`.

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildSettings.cs:
```diff
@@ -38,6 +38,12 @@ public ISet<string> Targets
+        /// Gets or sets the MSBuild target.
+        /// </summary>
+        /// <value>The MSBuild target.</value>
+        public MSBuildTarget? MSBuildTarget { get; set; }
```
Should not be nullable and should be explicitly set to `Automatic` in the constructor.

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs:
```diff
@@ -173,7 +173,23 @@ protected override IEnumerable<FilePath> GetAlternativeToolPaths(MSBuildSettings
             }
 
-            var path = MSBuildResolver.GetMSBuildPath(_fileSystem, _environment, settings.ToolVersion, settings.PlatformTarget ?? PlatformTarget.MSIL);
+            MSBuildTarget target = settings.MSBuildTarget ?? MSBuildTarget.Default;
```
Since `MSBuildTarget` shouldn't be nullable, there is no need for null coalescing here.

patriksvensson commented:
@marcosnz Yes, that would work. Left you some comments as well. :+1:

marcosnz commented:
Thanks @patriksvensson! I have incorporated your comments. How does it look now?

patriksvensson commented:
@marcosnz I realized that the tests should not be changed. The behavior should be the same after your change so the tests should remain what they looked like before your changes. To test your changes, we should add new tests that test the combination of platform target and msbuild platform.

Other than that it looks good to me :+1:

marcosnz commented:
@patriksvensson - of course! I have added back original tests, kept in my change tests that test the new property and added a couple of tests that have both properties set to ensure we are getting the results we expect there.

Redth commented:
To throw another variable into the mix, I believe that Windows Phone 81 projects that you want to compile for ARM must use the x86 msbuild.exe...   So you'd need to invoke it something like 

`...\x86\msbuild.exe /p:Configuration=Release /p:Platform=ARM /target:Build "SomeWp81ARM.sln"`

Maybe we need `MSBuildPlatformTarget.Arm`?

patriksvensson commented:
@Redth I agree ARM support sounds like a good idea but we should probably add an issue for it and do that on a separate PR and not in this one.

As a workaround, you can send a custom platform by doing this:

```csharp
MSBuild("./src/Cake.sln", new MSBuildSettings()
    .WithProperty("platform", "ARM")
```

marcosnz commented:
@Redth @patriksvensson I think that ARM would be added to PlatformTarget enum, not MSBuildPlatform?

Redth commented:
@marcosnz yes i do think that's correct now, the only thing we need to ensure is that the logic of which msbuild.exe to select can allow a combination of: `MSBuildPlatform.x86 + PlatformTarget.Arm`.

Redth commented:
@patriksvensson currently your suggested workaround won't work since for whatever reason microsoft requires the windows phone projects to be built with the x86 msbuild.exe when specifying `/p:platform=arm`, and right now the amd64 gets picked if running on a 64 bit system and PlatformTarget.x86 is not selected.

marcosnz commented:
@Redth @patriksvensson I have added support for ARM. I added it into this PR because of the use-case's dependency of the MSBuildPlatform. I added unit tests and if the MSBuildPlatform is automatic the ARM uses same logic as MSIL for choosing the version of MSBuild. I added a test case specifically for the `MSBuildPlatform.x86 + PlatformTarget.Arm` combination

patriksvensson commented:
Ok, this looks good to me.

Can you rebase to one commit (and after that against develop) and I'll merge it :+1:

marcosnz commented:
Umm... I think I did it? lol. My GIt knowledge is weak...

patriksvensson commented:
:+1: Perfect! :smile:

@devlead @gep13 Anyone of you have time to do a final review and merge this if you're happy with it? Otherwise I can do it in a couple of hours.

patriksvensson commented:
Merged :+1: 

I rebases the PR and updated the commit message to describe the changess in more details, that's why it doesn't appear as merged. Sorry about this, but didn't want to bother you (again) for a rebase.

Thank you very much for this! :smile:

patriksvensson commented:
Commit can be seen here: https://github.com/cake-build/cake/commit/144f6aa14e24ab1eb11499da4404dfb48c73d93b
